### PR TITLE
added OptType and optional ops

### PR DESCRIPTION
**Pull request by sorenlassen**
implemented shape inference for the ops

added lit tests

replaced isOptional logic from gen_onnx_mlir.py

note that this PR doesn't lower OptType (type conversion) and lower optional ops

### CODE REVIEW

**Comment by AlexandreEichenberger** [comment]
> TODO: lower OptType (type conversion) and lower optional ops

please use the "draft mode" so that we know when we need to start reviewing. You can always put it on draft and explicitly ask us to review it if you want early feedback.

Tx

**Comment by sorenlassen** [comment]
> > TODO: lower OptType (type conversion) and lower optional ops
> 
> please use the "draft mode" so that we know when we need to start reviewing. You can always put it on draft and explicitly ask us to review it if you want early feedback.

sorry for the confusion but I was hoping we could land this PR as is and save lowering for another round

I reworded "TODO" to "note that this PR doesn't lower" :)

BTW, it's not immediately clear to me how to do the type conversion for OptType to lower it to krnl, I couldn't see anything obvious in KrnlTypeConverter to pattern match against - any suggestions are welcome

**Comment by AlexandreEichenberger** [comment]
go it, thanks for the clarification. On our side, @chentong319 is the one that introduces the types, so I would recommend to exchange info with him. Tong is away for another week or so, he should be able to help once he is back.

File: utils/gen_onnx_mlir.py [hunk 1]
```diff
@@ -485,7 +491,7 @@ def onnx_attr_type_to_mlir_attr_type(t):
         mlir_attr_type = 'StrAttr'
     elif onnx_attr_type == "strings":
         mlir_attr_type = 'StrArrayAttr'
-    elif onnx_attr_type == 'type':
+    elif onnx_attr_type in {'type', 'type_proto'}:
```
Comment (AlexandreEichenberger) [inline utils/gen_onnx_mlir.py:494]: Can you just comment on what this new `type_proto` is for us not familiar with the onnx types, tx

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, just had to minor requests, would be great if you could answer them before dropping the code. Thanks for adding the Optional

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 2]
```diff
@@ -4417,6 +4417,53 @@ LogicalResult ONNXOneHotOp::inferShapes(
...
+}
+
+LogicalResult ONNXOptionalOp::inferShapes(
+    std::function<void(mlir::Region &)> doShapeInference) {
+  Type ty;
+  if (auto typeAttr = type()) {
+    ty = typeAttr.value();
+  } else {
+    ty = input().getType();
+    assert(!ty.isa<NoneType>());
```
[the remark this answers is not in the document]
Comment (sorenlassen) [reply src/Dialect/ONNX/ONNXOps.cpp]: done:
```
    // checked in verify()
    assert(!ty.isa<NoneType>() && "type attribute or input value needed");
```

File: utils/gen_onnx_mlir.py (hunk 1 above)
Comment (sorenlassen) [reply utils/gen_onnx_mlir.py:494]: added the comment:
```
        # 'type' is the attribute type used in special_attr_types,
        # 'type_proto' is Optional op's type attribute's type
```

**Pull request merged**